code/addons/onboarding/src/Onboarding.tsx

```diff
@@ -0,0 +1,276 @@
+    // @ts-ignore Ignore circular reference
+    Step,
+    | 'content'
+    | 'disableBeacon'
+    | 'disableOverlay'
+    | 'floaterProps'
+    | 'offset'
+    | 'placement'
+    | 'spotlightClicks'
+    | 'styles'
+    | 'target'
+    | 'title'
+  >
+>;
+
+export default function Onboarding({ api }: { api: API }) {
+  const [enabled, setEnabled] = useState(true);
+  const [showConfetti, setShowConfetti] = useState(false);
+  const [step, setStep] = useState<StepKey>('1:Intro');
+
+  const [primaryControl, setPrimaryControl] = useState<HTMLElement | null>();
+  const [saveFromControls, setSaveFromControls] = useState<HTMLElement | null>();
+  const [createNewStoryForm, setCreateNewStoryForm] = useState<HTMLElement | null>();
+  const [createdStory, setCreatedStory] = useState<{
+    newStoryName: string;
+    sourceFileContent: string;
+    sourceFileName: string;
+  } | null>();
+
+  const selectStory = useCallback(
+    (storyId: string) => {
+      try {
+        const { id, refId } = api.getCurrentStoryData() || {};
+        if (id !== storyId || refId !== undefined) api.selectStory(storyId);
+      } catch (e) {}
+    },
+    [api]
+  );
+
+  const disableOnboarding = useCallback(() => {
+    // remove onboarding query parameter from current url
+    const url = new URL(window.location.href);
+    // @ts-expect-error (not strict)
+    const path = decodeURIComponent(url.searchParams.get('path'));
+    url.search = `?path=${path}&onboarding=false`;
+    history.replaceState({}, '', url.href);
+    api.setQueryParams({ onboarding: 'false' });
+    setEnabled(false);
+  }, [api, setEnabled]);
+
+  const completeOnboarding = useCallback(() => {
+    api.emit(STORYBOOK_ADDON_ONBOARDING_CHANNEL, {
+      step: '6:FinishedOnboarding' satisfies StepKey,
+      type: 'telemetry',
+    });
+    selectStory('configure-your-project--docs');
+    disableOnboarding();
+  }, [api, selectStory, disableOnboarding]);
+
+  useEffect(() => {
+    api.setQueryParams({ onboarding: 'true' });
+    selectStory('example-button--primary');
+    api.togglePanel(true);
+    api.togglePanelPosition('bottom');
+    api.setSelectedPanel('addon-controls');
+  }, [api, selectStory]);
+
+  useEffect(() => {
+    const observer = new MutationObserver(() => {
+      setPrimaryControl(document.getElementById('control-primary'));
+      setSaveFromControls(document.getElementById('save-from-controls'));
+      setCreateNewStoryForm(document.getElementById('create-new-story-form'));
+    });
+
+    observer.observe(document.body, { childList: true, subtree: true });
+    return () => observer.disconnect();
+  }, []);
+
+  useEffect(() => {
+    setStep((current) => {
+      if (['1:Intro', '5:StoryCreated', '6:FinishedOnboarding'].includes(current)) return current;
+      if (createNewStoryForm) return '4:CreateStory';
+      if (saveFromControls) return '3:SaveFromControls';
+      if (primaryControl) return '2:Controls';
```

Comment:
We usually use block statements for if conditions. I think that's also something that @ndelangen prefers.

Reply:
In general, I would agree. In the context of this useEffect hook, I think adding a block for each if statement would only serve to add visual noise with no real benefit to legibility. The crux for me is that this is okay when it's consistent. In this case there's a block with 5 lines of code each ending with a return statement. That's consistent. If there would be other logic in this block as well, then it would not be consistent and I would've structured it differently.